src/lib/controllers/ui-controller.js

```diff
@@ -844,16 +846,24 @@ export default class UIController {
     this.store.commit(`ui/setNotification`, { text: message, important: true })
   }
 
+  /**
+   * Switched between tabs in a panel.
+   * All tab switching should be done through this function only as it performs safety check
+   * regarding wither or not current tab can be available.
+   * @param {string} tabName - A name of a tab to switch to.
+   * @return {UIController} - An instance of a UI controller, for chaining.
+   */
   changeTab (tabName) {
-    const statusAvailable = Boolean(this.api.settings.contentOptions.items.verboseMode.currentValue === 'verbose')
     // If tab is disabled, switch to a default one
     if (
-      /* (!this.store.state.app.tabState.hasOwnProperty(tabName)) || */
-      (!this.store.state.app.hasInflData && name === 'inflections') ||
-      (!this.store.getters['app/hasGrammarRes'] && name === 'grammar') ||
-      (!this.store.getters['app/hasTreebankData'] && name === 'treebank') ||
-      (!statusAvailable && name === 'status')
+      (name === 'definitions' && !this.store.getters['app/defDataReady']) ||
+      (name === 'inflections' && !this.store.state.app.hasInflData) ||
+      (name === 'grammar' && !this.store.getters['app/hasGrammarRes']) ||
+      (name === 'treebank' && !this.store.getters['app/hasTreebankData']) ||
+      (name === 'wordUsage' && !this.store.state.app.wordUsageExampleEnabled) ||
+      (name === 'status' && this.api.settings.contentOptions.items.verboseMode.currentValue !== 'verbose')
```

Comment:
May be it is better to convert it to Map, array or an object and iterate using a loop construction?

Reply:
I agree that this is verbose and a bit straightforward, but what I like about it is that it lists clearly what is a condition for each tab to be visible, so we can review it at a glance.

Do you have any ideas how to improve it without loosing cleanliness and adding too much unnecessary complexity?

Reply:
I meant to wrap it inside for example an object (as the order here is not important), for example:
```
let tabsCheck = {
  'grammar' : 'app/hasGrammarRes',
  'treebank': 'app/hasTreebankData'
  ...
  // other checks also could be converted to getters to make it more unified and easier for reading and understanding by others
}
 if (!this.store.getters[tabsCheck[tabName]] {
    console.warn ....
}
```

From my point of view it is clearly and easier for working with.

also, @kirlat , you are using a `name` variable inside a mehod, but it is not initialized not inside and not in arguments - are you sure, that it is working and this condition is reachable? Did you have tests for that?

Reply:
It's a very good catch with the `name`! It should be `tabName` instead. I fixed that, created a separate `comp-pr393` branch to implement suggestions from this PR.

We don't have up to date tests for the UI controller currently, but I'm not sure when it will be the best time to start working on that. @balmas, what's your opinion on that? 

I generally like the idea to make the tab check simpler to understand. I've tried to implement the simplest way I was able to find and I've got:
```
let tabsCheck = {
      definitions: () => this.store.getters['app/defDataReady'],
      inflections: () => this.store.state.app.hasInflData,
      grammar: () => this.store.getters['app/hasGrammarRes'],
      treebank: () => this.store.getters['app/hasTreebankData'],
      wordUsage: () => this.store.state.app.wordUsageExampleEnabled,
      status: () => this.api.settings.contentOptions.items.verboseMode.currentValue !== 'verbose'
    }
    if (tabsCheck.hasOwnProperty(tabName) && !tabsCheck[tabName]()) {
      console.warn(`Attempting to switch to a ${tabName} tab which is not available`)
      tabName = this.defaultTab
    }
```

vs. the old code which is:
```
if (
      (tabName === 'definitions' && !this.store.getters['app/defDataReady']) ||
      (tabName === 'inflections' && !this.store.state.app.hasInflData) ||
      (tabName === 'grammar' && !this.store.getters['app/hasGrammarRes']) ||
      (tabName === 'treebank' && !this.store.getters['app/hasTreebankData']) ||
      (tabName === 'wordUsage' && !this.store.state.app.wordUsageExampleEnabled) ||
      (tabName === 'status' && this.api.settings.contentOptions.items.verboseMode.currentValue !== 'verbose')
    ) {
      console.warn(`Attempting to switch to a ${tabName} tab which is not available`)
      tabName = this.defaultTab
    }
```

I think the new code is easier to read, even if it has little more characters in it 🙂. I'm wondering if it can be any simpler.

The check we do here is the same check we do in the panel when we decide whether to show any icon in a tab or not. So I think the real benefit would be if we combine both pieces of logic. But then in order to be shared and reactive (to benefit from caching) we have to put it to either Vuex or the panel module. @balmas, what do you think about that?

Reply:
I don't object to simplifying this.  As for the tests, I definitely do want us to have these tests but I think maybe that should wait until we reach the beta testing milestone, which I am working on scoping today.